var/spack/repos/builtin/packages/ldc/package.py

```diff
@@ -0,0 +1,84 @@
+##############################################################################
+# Copyright (c) 2013-2017, Lawrence Livermore National Security, LLC.
+# Produced at the Lawrence Livermore National Laboratory.
+#
+# This file is part of Spack.
+# Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
+# LLNL-CODE-647188
+#
+# For details, see https://github.com/llnl/spack
+# Please also see the NOTICE and LICENSE files for our notice and the LGPL.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License (as
+# published by the Free Software Foundation) version 2.1, February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+from spack import *
+
+import spack.environment
+import llnl.util.tty as tty
+
+
+class Ldc(CMakePackage):
+    """The LDC project aims to provide a portable D programming language
+    compiler with modern optimization and code generation capabilities.
+
+    LDC is fully Open Source; the parts of the code not taken/adapted from
+    other projects are BSD-licensed (see the LICENSE file for details).
+
+    Consult the D wiki for further information: http://wiki.dlang.org/LDC
+    """
+
+    homepage = "https://dlang.org/"
+    url = "https://github.com/ldc-developers/ldc/releases/download/v0.17.4/ldc-0.17.4-src.tar.gz"
+
+    version('1.3.0', '537d992a361b0fd0440b24a5145c9107')
+    # This is the last version that does not require a D compiler to bootstrap
+    version('0.17.4', '000e006426d6094fabd2a2bdab0ff0b7')
+
+    depends_on('llvm@3.7:')
+    depends_on('zlib')
+    depends_on('libconfig')
+    depends_on('curl')
+    depends_on('libedit')
+    depends_on('binutils')
+
+    def cmake_args(self):
+        args = [
+            '-DBUILD_SHARED_LIBS:BOOL=ON'
+        ]
+
+        if self.spec.satisfies('@1.0:'):
+            bootstrap = Spec('ldc@0.17.4')
+            bootstrap.concretize()
+
+            if not bootstrap.package.installed:
+                # ldc greater than 0.17.X needs a D compiler to be
+                # bootstrapped. There's no way currently to express
+                # this in Spack with directives, hence this logic.
+                tty.warn('ldc requires a D compiler to build.')
+                msg = 'Bootstrapping "{0}" right now'.format(
+                    bootstrap.cshort_spec
+                )
+                tty.warn(msg)
+                bootstrap.package.do_install()
```

Comment:
@adamjstewart @davydden @becker33 To the best of my knowledge, this is the first time in the `builtin` repository that any package tries to concretize and install another package. If there are no objection against it I would merge the package, otherwise I'll be happy to discuss any concern that this may raise.

Reply:
I think it's the usual case for a compiler written in the language it compiles. I think we'll see cases like that only for compilers of not widely used languages.

Reply:
The `go` package has a similar problem but solves it in a completely different way (by adding a `go-bootstrap` package with the older version). We should standardize how we want to handle this. I'm a little worried about your solution. It duplicates a lot of work, and every time a change happens to core, this package will further diverge from the standard installation process.

Reply:
> It duplicates a lot of work, and every time a change happens to core, this package will further diverge from the standard installation process.

Can you expand on this? I don't get where the work is duplicated and what could happen here that won't happen in `go` + `go-bootstrap` as core changes.

Reply:
Well, `concretize`, `do_install`, and all of the environment modifications are all duplicated from core. Does your process even run post-installation hooks? Every time core changes things like that, this package won't be updated. At least with `go-boostrap`, the only things that are duplicated are pretty simple things like a docstring, homepage, and installation flags. Those don't worry me as much as your technique.

Reply:
@adamjstewart As far as I can tell I am not *duplicating* anything. The only thing that changes is that I'll call two times `setup-environment` and spawn an additional process (because, if not present, I'll install a package during a package installation). On the other hand I am not creating a non-existing package that only serves the purpose of bootstrapping.

That said I am not opposed to extract the old compiler version into a `ldc-bootstrap` package if people think it's safer. I just don't get where I am duplicating stuff here.

Reply:
You are duplicating calls to `concretize`, `package.do_install`, and `EnvironmentModifications`. These commands are already present in core. Basically, you are copying a trimmed-down version of how packages get installed into a package itself.

Reply:
> You are duplicating calls to concretize, package.do_install, and EnvironmentModifications

Ah, ok. It's true that I am concretizing a package within another package (that's why I asked for opinions in the first place). But calling a public API is not **duplicating code** - that was the point I didn't understand.

So I'll be waiting for other opinions. For now +1 to separate `ldc-bootstrap`.

Reply:
I like the idea of consistency so I'm in favor of doing it as go does.  Perhaps once we get to 3-ish examples there might be enough info for a generalized solution.

Reply:
@hartzell I missed your comment here, so +2 for `ldc-bootstrap`. I think that's it - I'll separate the package. @tgamblin @scheibelp Sorry for the noise with the message below.